### PR TITLE
Optimize endpoint index getter and also rename some methods to be mor…

### LINKED ISSUE #775: Enhance workload deletion processing
**What would you like to be added**:
Enhance workload deletion processing
**Why is this needed**:
When the workload is deleted, you need to find the entire endpoint map, which slows down the performance of kmesh.
https://github.com/kmesh-net/kmesh/blob/e85c3d10c65181057e942e2e2c386964b4de4741/pkg/controller/workload/bpfcache/endpoint.go#L47-L63
When Kmesh used on a large scale, this will severely slow down the performances.

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
…e readable

**What type of PR is this?**

/kind cleanup
/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #775

Previously we iterate all the endpoint map when delete a workload. Now the index are cached in userspace

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @LiZhenCheng9527

File: pkg/controller/workload/bpfcache/endpoint.go [hunk 1]
```diff
@@ -31,33 +32,81 @@ type EndpointValue struct {
...
+
 	return c.bpfMap.KmeshEndpoint.Delete(key)
 }
 
-func (c *Cache) EndpointLookup(key *EndpointKey, value *EndpointValue) error {
-	log.Debugf("EndpointLookup [%#v]", *key)
-	return c.bpfMap.KmeshEndpoint.Lookup(key, value)
-}
+// EndpointSwap update the last endpoint index and remove the current endpoint
+func (c *Cache) EndpointSwap(currentIndex, lastIndex uint32, serviceId uint32) error {
```
Comment (LiZhenCheng9527) [inline pkg/controller/workload/bpfcache/endpoint.go:62]: Do you need to add a debug log like other function?

File: pkg/controller/workload/workload_processor.go [hunk 2]
```diff
@@ -272,54 +278,48 @@ func (p *Processor) storeEndpointWithService(sk *bpf.ServiceKey, sv *bpf.Service
...
+func (p *Processor) handleWorkloadUnboundServices(workload *workloadapi.Workload, services []string) error {
+	if len(services) == 0 {
 		return nil
 	}
 
-	log.Infof("deleteResidualServicesWithWorkload: %v", services)
-	eks := make([]bpf.EndpointKey, 0)
-	workloadUid := p.hashName.StrToNum(workload.GetUid())
-	serviceIds := make(map[uint32]struct{})
+	log.Infof("handleWorkloadUnboundServices %s: %v", workload.Uid, services)
```
Comment (LiZhenCheng9527) [inline pkg/controller/workload/workload_processor.go:301]: Can replace workload.Uid with workload.name?

File: pkg/controller/workload/workload_processor.go [hunk 3]
```diff
@@ -330,14 +330,14 @@ func (p *Processor) addNewServicesWithWorkload(workload *workloadapi.Workload, n
 		return nil
 	}
 
-	log.Infof("addNewServicesWithWorkload: %v", newServices)
-	backend_uid := p.hashName.StrToNum(workload.GetUid())
+	log.Infof("handleWorkloadNewBoundServices %s: %v", workload.Uid, newServices)
```
Comment (LiZhenCheng9527) [inline pkg/controller/workload/workload_processor.go:333]: ditto

**Review by LiZhenCheng9527 — COMMENTED** [review]
/lgtm

File: pkg/controller/workload/workload_processor.go [hunk 4]
```diff
@@ -160,24 +164,25 @@ func (p *Processor) removeWorkloadFromBpfMap(uid string) error {
...
 		if err != nil {
 			return err
 		}
 	}
 
+	// 3. delete workload from backend map
 	bkDelete.BackendUid = backendUid
 	if err = p.bpf.BackendDelete(&bkDelete); err != nil {
-		log.Errorf("BackendDelete failed: %s", err)
+		log.Errorf("BackendDelete failed %v: %s", backendUid, err)
```
Comment (kwb0523) [inline pkg/controller/workload/workload_processor.go:185]:
```suggestion
		log.Errorf("BackendDelete %v failed: %s", backendUid, err)
```

File: pkg/controller/workload/workload_hash.go [hunk 5]
```diff
@@ -90,7 +90,7 @@ func (h *HashName) flushDelta(str string, num uint32) error {
 	return err
 }
 
-func (h *HashName) StrToNum(str string) uint32 {
+func (h *HashName) Hash(str string) uint32 {
```
Comment (kwb0523) [inline pkg/controller/workload/workload_hash.go:93]: The function name seems a bit abstract and easy to confuse.

Comment (hzxuzhonghu) [reply pkg/controller/workload/workload_hash.go:93]: on the contrary, i think this is consistent with go hash  pkg

File: pkg/controller/workload/workload_processor.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/controller/workload/workload_processor.go:301]: isn't Uid contains more info? like namespace and kind info

File: pkg/controller/workload/workload_processor.go [hunk 6]
```diff
@@ -330,14 +336,14 @@ func (p *Processor) addNewServicesWithWorkload(workload *workloadapi.Workload, n
...
+	workloadId := p.hashName.Hash(workload.GetUid())
 	for _, serviceName := range newServices {
-		sk.ServiceId = p.hashName.StrToNum(serviceName)
+		sk.ServiceId = p.hashName.Hash(serviceName)
 		// the service already stored in map, add endpoint
 		if err = p.bpf.ServiceLookup(&sk, &sv); err == nil {
-			if err = p.storeEndpointWithService(&sk, &sv, backend_uid); err != nil {
-				log.Errorf("storeEndpointWithService failed, err:%s", err)
+			if err = p.addWorkloadToService(&sk, &sv, workloadId); err != nil {
+				log.Errorf("addWorkloadToService failed, err:%s", err)
```
Comment (YaoZengzeng) [inline pkg/controller/workload/workload_processor.go:346]:
```suggestion
				log.Errorf("addWorkloadToService failed: %v", err)
```

File: pkg/controller/workload/workload_processor.go [hunk 7]
```diff
@@ -400,19 +404,19 @@ func (p *Processor) handleWorkload(workload *workloadapi.Workload) error {
...
+	if err := p.handleWorkloadUnboundServices(workload, deletedServices); err != nil {
+		log.Errorf("handleWorkloadUnboundServices %s failed: %v", workload.GetUid(), err)
 		return err
 	}
 
 	// Add new services associated with the workload
-	if err := p.addNewServicesWithWorkload(workload, newServices); err != nil {
-		log.Errorf("addNewServicesWithWorkload %s failed: %v", workload.Uid, err)
+	if err := p.handleWorkloadNewBoundServices(workload, newServices); err != nil {
+		log.Errorf("handleWorkloadNewBoundServices %s failed: %v", workload.Uid, err)
```
Comment (YaoZengzeng) [inline pkg/controller/workload/workload_processor.go:415]:
```suggestion
		log.Errorf("handleWorkloadNewBoundServices %s failed: %v", workload.GetUid(), err)
```

File: pkg/controller/workload/workload_processor.go [hunk 8]
```diff
@@ -503,8 +506,19 @@ func (p *Processor) storeServiceData(serviceName string, waypoint *workloadapi.G
 	return nil
 }
 
+func (p *Processor) shouldAddEndpoint(wlUid uint32, serviceUid uint32) bool {
+	eks := p.bpf.GetEndpointKeys(wlUid)
+	for k := range eks {
+		if k.ServiceId == serviceUid {
+			log.Debugf("workload %d has been storred as endpoint of service %d", wlUid, serviceUid)
```
Comment (YaoZengzeng) [inline pkg/controller/workload/workload_processor.go:513]:
```suggestion
			log.Debugf("workload %d has been stored as endpoint of service %d", wlUid, serviceUid)
```

**Pull request merged**